broadcastclients/broadcastclients.go

```diff
@@ -72,12 +121,76 @@ func (bcs *BroadcastClients) adjustCount(delta int32) {
 }
 
 func (bcs *BroadcastClients) Start(ctx context.Context) {
-	for _, client := range bcs.clients {
+	bcs.router.StopWaiter.Start(ctx, bcs.router)
+
+	for _, client := range bcs.primaryClients {
 		client.Start(ctx)
 	}
+
+	var lastConfirmed arbutil.MessageIndex
+	recentFeedItemsNew := make(map[arbutil.MessageIndex]time.Time, RECENT_FEED_INITIAL_MAP_SIZE)
+	recentFeedItemsOld := make(map[arbutil.MessageIndex]time.Time, RECENT_FEED_INITIAL_MAP_SIZE)
+	bcs.router.LaunchThread(func(ctx context.Context) {
+		recentFeedItemsCleanup := time.NewTicker(RECENT_FEED_ITEM_TTL)
+		startNewFeedTimer := time.NewTicker(MAX_FEED_INACTIVE_TIME)
+		defer recentFeedItemsCleanup.Stop()
+		defer startNewFeedTimer.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case msg := <-bcs.router.messageChan:
+				startNewFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+				if _, ok := recentFeedItemsNew[msg.SequenceNumber]; ok {
+					continue
+				}
+				if _, ok := recentFeedItemsOld[msg.SequenceNumber]; ok {
+					continue
+				}
+				recentFeedItemsNew[msg.SequenceNumber] = time.Now()
+				// need to stop the timer because forwardTxStreamer might be blocked when traffic is high
+				// and that shouldn't create race condition between channels timer.C and messageChan
+				startNewFeedTimer.Stop()
+				if err := bcs.router.forwardTxStreamer.AddBroadcastMessages([]*broadcaster.BroadcastFeedMessage{&msg}); err != nil {
+					log.Error("Error routing message from Sequencer Feed", "err", err)
+				}
+				startNewFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+			case cs := <-bcs.router.confirmedSequenceNumberChan:
+				startNewFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+				if cs == lastConfirmed {
+					continue
+				}
+				lastConfirmed = cs
+				startNewFeedTimer.Stop()
+				bcs.router.forwardConfirmationChan <- cs
+				startNewFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+			case <-recentFeedItemsCleanup.C:
+				// Cycle buckets to get rid of old entries
+				recentFeedItemsOld = recentFeedItemsNew
+				recentFeedItemsNew = make(map[arbutil.MessageIndex]time.Time, RECENT_FEED_INITIAL_MAP_SIZE)
+			case <-startNewFeedTimer.C:
+				// failed to get messages from primary feed for ~5 seconds, start a new feed
+				bcs.StartSecondaryFeed(ctx)
+			}
+		}
+	})
 }
+
+func (bcs *BroadcastClients) StartSecondaryFeed(ctx context.Context) {
+	if bcs.numOfStartedSecondary < len(bcs.secondaryClients) {
+		client := bcs.secondaryClients[bcs.numOfStartedSecondary]
+		bcs.numOfStartedSecondary += 1
+		client.Start(ctx)
```

Comment:
I think we still want a way to stop clients that are no longer needed anymore, otherwise a temporary network blip will double the amount of relay traffic if all external nitro instances are using the same configuration with one primary and one secondary, until those nitro instances are restarted.

Reply:
Redirecting to @PlasmaPower

I think the issue was we would get stuck in the loop of disconnecting and reconnecting secondaries.
Maybe a good middle ground would be to have two routers one of primary feeds and the other for secondaries to measure traffic and disconnect secondaries if needed, in the scenarios like you mentioned.

Reply:
I think you could avoid that if you had some time where you would only disconnect from the secondary if the other feeds had no errors.